src/main/scala/is/hail/distributedmatrix/GridPartitioner.scala

```diff
@@ -2,47 +2,40 @@ package is.hail.distributedmatrix
 
 import org.apache.spark.Partitioner
 
-case class GridPartitioner(blockSize: Int, rows: Long, cols: Long) extends Partitioner {
-  require(rows > 0)
-  require(cols > 0)
-  require((rows - 1) / blockSize + 1 < Int.MaxValue)
-  require((cols - 1) / blockSize + 1 < Int.MaxValue)
-
-  val rowPartitions: Int = ((rows - 1) / blockSize + 1).toInt
-  val colPartitions: Int = ((cols - 1) / blockSize + 1).toInt
-  
-  val truncatedBlockRow: Int = (rows / blockSize).toInt
-  val truncatedBlockCol: Int = (cols / blockSize).toInt
-  
-  val excessRows: Int = (rows % blockSize).toInt
-  val excessCols: Int = (cols % blockSize).toInt
-  
-  def rowPartitionRows(i: Int): Int = if (i != truncatedBlockRow) blockSize else excessRows
-  def colPartitionCols(j: Int): Int = if (j != truncatedBlockCol) blockSize else excessCols
-  
-  override val numPartitions: Int = rowPartitions * colPartitions
-  assert(numPartitions >= rowPartitions && numPartitions >= colPartitions)
+case class GridPartitioner(blockSize: Int, nRows: Long, nCols: Long) extends Partitioner {
+  require(nRows > 0)
+  require(nCols > 0)
+  require((nRows - 1) / blockSize + 1 < Int.MaxValue)
+  require((nCols - 1) / blockSize + 1 < Int.MaxValue)
+
+  val nBlockRows: Int = ((nRows - 1) / blockSize + 1).toInt
+  val nBlockCols: Int = ((nCols - 1) / blockSize + 1).toInt
   
+  def blockRowNRows(i: Int): Int =
+    (if (i < nBlockRows - 1) blockSize else nRows - (nBlockRows - 1) * blockSize).toInt
```

Comment:
I'd move the left `(` here and in blockColNCols:
`if (i < nBlockRows - 1) blockSize else (nRows - (nBlockRows - 1) * blockSize).toInt